src/basilisp/core.lpy

```diff
@@ -4437,16 +4437,31 @@
   *default-data-reader-fn*
   nil)
 
-(defn- read-iterator
-  [opts x]
-  (let [read (:read opts basilisp.lang.reader/read)]
-    (read x
-          *resolver*
-          *data-readers*
-          (:eof opts)
-          (= (:eof opts) :eofthrow)
-          (:features opts)
-          (not= :preserve (:read-cond opts)))))
+(defn read-seq
+  "Create an lazy sequence that will contain each form from the
+  ``stream``\\. If no stream is specified, uses the value currently
+  bound to :lpy:var:`*in*`. It is the callers responsibility to ensure
+  that the stream resource is not closed before the sequence has been
+  realised.
+
+  Callers may bind a map of readers to :lpy:var:`*data-readers*` to customize
+  the data readers used reading this string
+
+  The stream must satisfy the interface of :external:py:class:`io.TextIOBase`\\, but
+  does not require any pushback capabilities. The default
+  ``basilisp.lang.reader.StreamReader`` can wrap any object implementing ``TextIOBase``
+  and provide pushback capabilities."
+  ([stream]
+   (read-seq {} stream))
+  ([opts stream]
+   (let [read (:read opts basilisp.lang.reader/read)]
+      (lazy-seq (read stream
+                      *resolver*
+                      *data-readers*
+                      (:eof opts)
+                      (= (:eof opts) :eofthrow)
+                      (:features opts)
+                      (not= :preserve (:read-cond opts)))))))
```

Comment:
`read` is already lazy (in the Python sense). What value is this adding?

Reply:
I filed https://github.com/basilisp-lang/basilisp/issues/1005 to address this separately.

Reply:
`(seq (read ...))` works and will yield a lazy sequence for Python generator types. It is more correct than wrapping in `lazy-seq` because it closes over the generator iterator and ensures that only that single LazySeq can call `next`.

I'm increasingly convinced there's no bug at all and this is simply the cost of interacting with Python generators, but am still investigating.